src/cluster_asm.c

```diff
@@ -2323,3 +2331,176 @@ int clusterAsmProcess(const char *task_id, int event, void *arg, char **err) {
 
     return ret;
 }
+
+/* Check if we can propagate TRIMSLOTS command to AOF and replicas. */
+static int canPropagateTrimSlots(void) {
+    return !isPausedActions(PAUSE_ACTION_CLIENT_WRITE) &&
+           !isPausedActions(PAUSE_ACTION_CLIENT_ALL) &&
+           !isPausedActions(PAUSE_ACTION_REPLICA);
+}
+
+/* Propagate TRIMSLOTS command to AOF and replicas. */
+static void propagateTrimSlots(slotRangeArray *slots) {
+    int argc = slots->num_ranges * 2 + 3;
+    robj **argv = zmalloc(sizeof(robj*) * argc);
+    argv[0] = createStringObject("TRIMSLOTS", 9);
+    argv[1] = createStringObject("RANGES", 6);
+    argv[2] = createStringObjectFromLongLong(slots->num_ranges);
+    for (int i = 0; i < slots->num_ranges; i++) {
+        argv[i*2+3] = createStringObjectFromLongLong(slots->ranges[i].start);
+        argv[i*2+4] = createStringObjectFromLongLong(slots->ranges[i].end);
+    }
+
+    enterExecutionUnit(1, 0);
+
+    int prev_replication_allowed = server.replication_allowed;
+    server.replication_allowed = 1;
+    alsoPropagate(-1, argv, argc, PROPAGATE_AOF | PROPAGATE_REPL);
+    server.replication_allowed = prev_replication_allowed;
+
+    exitExecutionUnit();
+    postExecutionUnitOperations();
+
+    for (int i = 0; i < argc; i++)
+        decrRefCount(argv[i]);
+    zfree(argv);
+}
+
+/* Trim the slots asynchronously in the BIO thread. */
+void asmTrimInBackground(slotRangeArray *slots) {
+    /* TODO: Fire module event
+     * TODO: This is going to send invalidation message for all the keys for
+     * the tracking clients. We need to consider how we can do that for only
+     * the keys in the slots we are trimming. */
+    signalFlushedDb(0, 1, slots);
+
+    /* Create temp kvstores, move relevant slot dicts into them, and delete them
+     * in BIO thread asynchronously. */
+    kvstore *keys = kvstoreCreate(&dbDictType,
+                                  CLUSTER_SLOT_MASK_BITS,
+                                  KVSTORE_ALLOCATE_DICTS_ON_DEMAND);
+    kvstore *expires = kvstoreCreate(&dbExpiresDictType,
+                                     CLUSTER_SLOT_MASK_BITS,
+                                     KVSTORE_ALLOCATE_DICTS_ON_DEMAND);
+
+    for (int i = 0; i < slots->num_ranges; i++) {
+        for (int slot = slots->ranges[i].start; slot <= slots->ranges[i].end; slot++) {
+            kvstoreMoveDict(server.db[0].keys, keys, slot);
+            kvstoreMoveDict(server.db[0].expires, expires, slot);
+            /* TODO: hexpires */
+        }
+    }
+
+    /* TODO: Do not delete all the hexpires */
+    emptyDbDataAsync(keys, expires, server.db[0].hexpires);
```

Comment:
i remember you said we can trim in blocking method when loading

Reply:
yes, I have that logic as part of active trim PR. For bgtrim, it is okay to run in background similar to flushdb.  So, perhaps we can address that in the follow up PR.

Reply:
sounds good